### PR TITLE
Hide data constructors and uphold identity laws.

### PR DESCRIPTION
**Pull request by vreuter**
- This changes the way tests are done so that `ContentStore.Item` and `ContentHash` constructors can be re-hidden as before, per https://github.com/tweag/funflow2/pull/88#pullrequestreview-664491488.
- This removes de-duplication of `VolumeBindings` when `DockerTaskInstance`s are combined through the semigroup instance, so that identity laws hold. The implementation changes from a fold that selectively adds items in succession, to one that simply concatenates lists.

@dorranh are you OK with the semigroup behavior? The possibility of identical `VolumeBinding`s seems unlikely (and perhaps not even problematic?), so it felt OK to go w/ list concat but LMK what you think.

### CODE REVIEW

**Comment by vreuter** [comment]
To obviate need to expose the data constructors, used a similar task construct as in `TestFunflow`, but here mainly to get ahold of a `CS.Item`, which solved both the `Item` constructor use and the `ContentHash` one, since I'd exposed `ContentHash` specifically to make an `Item`

File: funflow/test/unit/Unit.hs [hunk 1]
```diff
@@ -1,21 +1,61 @@
...
+prop_DockerTaskInputMonoidRightIdentity = checkOneInput (\dti -> dti == dti <> mempty)
 
-prop_DockerTaskInputMonoidRightIdentity :: DockerTaskInput -> Bool
-prop_DockerTaskInputMonoidRightIdentity x = x <> mempty == x
+prop_DockerTaskInputMonoidAssociativity :: Property
+prop_DockerTaskInputMonoidAssociativity = QCM.monadicIO $ do
+    x <- buildDockerTaskInput
+    y <- buildDockerTaskInput
+    z <- buildDockerTaskInput
+    return ((x <> y) <> z == x <> (y <> z))
```
Comment (dorranh) [inline funflow/test/unit/Unit.hs:54]: :+1:

File: funflow/test/unit/Unit.hs [hunk 2]
```diff
@@ -1,21 +1,61 @@
...
+    }
+
+checkOneInput :: (DT.DockerTaskInput -> Bool) -> Property
+checkOneInput testPred = QCM.monadicIO $ testPred <$> buildDockerTaskInput
+
+prop_DockerTaskInputMonoidLeftIdentity :: Property
+prop_DockerTaskInputMonoidLeftIdentity = checkOneInput (\dti -> mempty <> dti == dti)
+
+prop_DockerTaskInputMonoidRightIdentity :: Property
+prop_DockerTaskInputMonoidRightIdentity = checkOneInput (\dti -> dti == dti <> mempty)
```
Comment (dorranh) [inline funflow/test/unit/Unit.hs:47]: :+1:

File: funflow/src/Funflow/Tasks/Docker.hs [hunk 3]
```diff
@@ -69,10 +67,13 @@ data DockerTaskInput = DockerTaskInput
 
 instance Semigroup DockerTaskInput where
   DockerTaskInput{ inputBindings = vols1, argsVals = args1 } <> DockerTaskInput{ inputBindings = vols2, argsVals = args2} = 
-    let agg (seen, xs) x = if Set.member x seen then (seen, xs) else (Set.insert x seen, x:xs)
+    -- TODO: deduplicate? If so, this becomes lawless unless, e.g., the collection of bindings is sth like Set.
+    {-let agg (seen, xs) x = if Set.member x seen then (seen, xs) else (Set.insert x seen, x:xs)
```
Comment (dorranh) [inline funflow/src/Funflow/Tasks/Docker.hs:71]: You know, this is actually a good point I hadn't thought about when we were discussing this earlier. 

> docker run -v $PWD/dummy:/mnt1 -v $PWD/dummy:/mnt2 -v $PWD/test:/mnt1  python:latest
docker: Error response from daemon: Duplicate mount point: /mnt1.
See 'docker run --help'.

It looks like you can not have duplicate mount points (though you can have volume bindings which point to the same host directory). In this case a better model for the volume bindings would be to store them in something like a map where the key is the mount point. That, unfortunately, will require quite a few changes throughout the codebase (and I would advise against spending time on that at the moment).

While inefficient, the simplest solution would probably be to handle this logic within `<>` for now.

**Review by dorranh — COMMENTED** [review]
Nice! Just a couple of comments. Also just to nitpick depending on how much time you have, it would be good to avoid calling `runFlow` in this test since the test is really focused on just the docker task config. Ideally we could construct a content store directly (without using the `StoreFlow` logic). (`tasty` also has support for QuickCheck if you want to use the `withResource` functionality there). https://hackage.haskell.org/package/tasty-quickcheck-0.10.1.2/docs/Test-Tasty-QuickCheck.html

File: funflow/src/Funflow/Tasks/Docker.hs (hunk 3 above)
Comment (vreuter) [reply funflow/src/Funflow/Tasks/Docker.hs:71]: OK so maybe best course is to...
- have the semigroup and monoid be "outlaws" for now
- open ticket to consider change to set/map like structure for the bindings?

**Pull request merged**